builder/build_unix_test.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) OpenFaaS Project 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+
+package builder
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"testing"
+)
+
+func Test_cp(t *testing.T) {
+	srcDir, srcDirErr := setupSourceFolder(2, []int{0600, 0640, 0644, 0700, 0755})
+	if srcDirErr != nil {
+		t.Error("Error creating source folder")
+	}
+	defer os.RemoveAll(srcDir)
+
+	destDir, destDirErr := ioutil.TempDir(os.TempDir(), "openfaas-test-destination-")
+	if destDirErr != nil {
+		t.Error("Error creating destination folder\n%v", destDirErr)
+	}
+	defer os.RemoveAll(destDir)
+
+	if files, err := ioutil.ReadDir(srcDir); err != nil {
+		t.Error("Cannot stat dir")
+	} else {
+		for _, f := range files {
+			if err := cp(srcDir+"/"+f.Name(), destDir+"/"+f.Name()); err != nil {
```

Comment:
Would it just be worth testing a single file with two different permissions to verify this behaviour? I think it would make the test much simpler.